File: app/js/engine.js
// engine.js — Moteur de tournoi du Dilemme du Prisonnier Itéré (Axelrod)
// --------------------------------------------------------------------------
// Matrice de gain (T,R,P,S) = (5,3,1,0)  — conforme à TourExec/AxTest.f
//   C/C -> R=3 / R=3      (coopération mutuelle)
//   D/D -> P=1 / P=1      (trahison mutuelle)
//   D/C -> T=5 / S=0      (tentation / pigeon)
// Encodage des coups : 0 = Coopérer, 1 = Trahir.
// Convention "pas de coup précédent" au tour 1 : opponentLastMove = -1.

export const COOPERATE = 0;
export const DEFECT = 1;
export const LENGTHS = [63, 77, 151, 156, 308];
export const PAYOFF = {
  '00': [3, 3],
  '11': [1, 1],
  '10': [5, 0], // joueur trahit (1), adv coopère (0) -> T/S
  '01': [0, 5], // joueur coopère (0), adv trahit (1) -> S/T
};

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

// Force toute valeur retournée par une stratégie vers 0 ou 1.
// Toute valeur invalide -> coup forfait (trahir) + marque une faute.
export function sanitizeMove(mv) {
  if (mv === 0 || mv === 1) return { move: mv, fault: false };
  if (mv === true) return { move: 1, fault: false };
  if (mv === false) return { move: 0, fault: false };
  const n = Number(mv);
  if (n === 0) return { move: 0, fault: false };
  if (n === 1) return { move: 1, fault: false };
  return { move: DEFECT, fault: true }; // coup forfait
}

// Une "stratégie" est un objet : { init(): instance, decide(instance, ctx): 0|1 }
// ctx = { opponentLastMove, currentTurn, myScore, opponentScore, randomValue, myLastMove }
// decide peut être synchrone ou asynchrone (Pyodide).

// Joue un match unique de `length` tours entre stratA et stratB.
// opts = { rng, noise (0..1), onTurn(event), delayMs, signal }
export async function playMatch(stratA, stratB, length, opts = {}) {
  const rng = opts.rng || Math.random;
  const noise = opts.noise || 0;
  const instA = await stratA.init();
  const instB = await stratB.init();
  let lastA = -1;
  let lastB = -1;
  let scoreA = 0;
  let scoreB = 0;
  let faultsA = 0;
  let faultsB = 0;
  const log = [];

  for (let turn = 1; turn <= length; turn++) {
    if (opts.signal?.aborted) throw new Error('aborted');

    const ctxA = {
      opponentLastMove: lastB,
      currentTurn: turn,
      myScore: scoreA,
      opponentScore: scoreB,
      randomValue: rng(),
      myLastMove: lastA,
    };
    const ctxB = {
      opponentLastMove: lastA,
      currentTurn: turn,
      myScore: scoreB,
      opponentScore: scoreA,
      randomValue: rng(),
      myLastMove: lastB,
    };

    let rawA = await stratA.decide(instA, ctxA);
    let rawB = await stratB.decide(instB, ctxB);

    let sA = sanitizeMove(rawA);
    let sB = sanitizeMove(rawB);
    if (sA.fault) faultsA++;
    if (sB.fault) faultsB++;
    let mA = sA.move;
    let mB = sB.move;

    // Bruit (trembling hand) : un coup est inversé avec probabilité `noise`.
    if (noise > 0) {
      if (rng() < noise) mA = 1 - mA;
      if (rng() < noise) mB = 1 - mB;
    }

    const [gainA, gainB] = PAYOFF[`${mA}${mB}`];
    scoreA += gainA;
    scoreB += gainB;

    const ev = {
      turn, mA, mB, gainA, gainB, scoreA, scoreB,
      rawA, rawB, faultA: sA.fault, faultB: sB.fault,
    };
    log.push(ev);
    lastA = mA;
    lastB = mB;

    if (opts.onTurn) opts.onTurn(ev, { instA, instB });
    if (opts.delayMs && opts.delayMs > 0) await sleep(opts.delayMs);
  }

  return {
    scoreA, scoreB, faultsA, faultsB, log, length,
    winner: scoreA > scoreB ? 'A' : scoreB > scoreA ? 'B' : 'DRAW',
  };
}

// Benchmark complet : 5 longueurs × N répétitions, moyenne des scores.
// Stratégies : re-instanciées à chaque match (joue aussi contre elle-même
// si a et b sont le même objet, pour mesurer la robustesse face à un clone).
export async function benchmark(stratA, stratB, { lengths = LENGTHS, reps = 10, rng, noise = 0, onProgress } = {}) {
  const results = [];
  let totalA = 0, totalB = 0, n = 0;
  for (const len of lengths) {
    let lenA = 0, lenB = 0;
    for (let r = 0; r < reps; r++) {
      const res = await playMatch(stratA, stratB, len, { rng, noise });
      lenA += res.scoreA;
      lenB += res.scoreB;
      totalA += res.scoreA;
      totalB += res.scoreB;
      n++;
      if (onProgress) onProgress({ len, rep: r + 1, reps, totalA, totalB, n });
    }
    results.push({ length: len, avgA: lenA / reps, avgB: lenB / reps });
  }
  return {
    results,
    avgA: totalA / n,
    avgB: totalB / n,
    matches: n,
  };
}

// Tournoi round-robin : chaque stratégie affronte toutes les autres ET elle-même
// (self-play inclus, à la Axelrod), sur les 5 longueurs de manches. Renvoie le
// classement du premier au dernier, trié par score moyen par tour (métrique
// equitable quelle que soit la longueur des manches).
//
// strategies : tableau d'objets { meta:{id,name}, init, decide }.
// opts : { lengths=LENGTHS, reps=1, onProgress(done, total, stats), onResult(i,j,length,res) }
//   onResult est appelé après chaque match (utile pour accumuler un head-to-head).
// Retour : tableau trié de { index, name, isUser, rank, totalScore, totalTurns,
//   matches, avgPerTurn, avgPerMatch, wins, ties, losses }.
export async function roundRobin(strategies, { lengths = LENGTHS, reps = 1, onProgress, onResult } = {}) {
  const n = strategies.length;
  const stats = strategies.map((s, i) => ({
    index: i,
    name: s.meta?.name || `Stratégie ${i + 1}`,
    isUser: !!s.meta?.isUser,
    totalScore: 0,
    totalTurns: 0,
    matches: 0,
    wins: 0,
    ties: 0,
    losses: 0,
  }));

  const totalPairs = (n * (n + 1)) / 2; // i <= j, self-play inclus
  let done = 0;

  for (let i = 0; i < n; i++) {
    for (let j = i; j < n; j++) {
      for (const len of lengths) {
        for (let r = 0; r < reps; r++) {
          const res = await playMatch(strategies[i], strategies[j], len);
          stats[i].totalScore += res.scoreA;
          stats[i].totalTurns += len;
          stats[i].matches += 1;
          stats[j].totalScore += res.scoreB;
          stats[j].totalTurns += len;
          stats[j].matches += 1;
          if (onResult) onResult(i, j, len, res);
          if (i !== j) {
            if (res.scoreA > res.scoreB) { stats[i].wins++; stats[j].losses++; }
            else if (res.scoreA < res.scoreB) { stats[i].losses++; stats[j].wins++; }
            else { stats[i].ties++; stats[j].ties++; }
          }
        }
      }
      done++;
      if (onProgress) onProgress(done, totalPairs, stats);
    }
  }

  stats.forEach((s) => {
    s.avgPerTurn = s.totalTurns ? s.totalScore / s.totalTurns : 0;
    s.avgPerMatch = s.matches ? s.totalScore / s.matches : 0;
  });
  stats.sort((a, b) => b.avgPerTurn - a.avgPerTurn || b.totalScore - a.totalScore);
  stats.forEach((s, i) => { s.rank = i + 1; });
  return stats;
}

File: backend/PdArena/Engine.cs
// Engine.cs — Moteur de tournoi : match unique + round-robin (Axelrod).
// --------------------------------------------------------------------------
// Port fidèle de app/js/engine.js.  Round-robin avec self-play (i <= j),
// sur les 5 longueurs de manches.  Classement par score moyen par tour
// (métrique équitable quelle que soit la longueur des manches).
namespace PdArena;

using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;

public static class Engine
{
    /// <summary>Force toute valeur retournée vers 0 ou 1.  Invalide -> forfait (trahir).</summary>
    public static (int move, bool fault) SanitizeMove(int mv)
    {
        if (mv == 0 || mv == 1) return (mv, false);
        return (Pd.Defect, true);
    }

    /// <summary>Joue un match de <paramref name="length"/> tours entre deux stratégies.</summary>
    /// <param name="turnTimeoutMs">Temps réel max par tour (0 = illimité).  Un dépassement
    /// vaut coup forfait (trahir) + faute — filet de sécurité anti-boucle infinie.</param>
    public static MatchResult PlayMatch(
        IPdStrategy stratA, IPdStrategy stratB, int length,
        Random rng, int turnTimeoutMs = 150,
        Action<int, int, int>? onTurn = null)
    {
        var instA = stratA.CreateInstance();
        var instB = stratB.CreateInstance();
        int lastA = Pd.NoMove, lastB = Pd.NoMove;
        int scoreA = 0, scoreB = 0, faultsA = 0, faultsB = 0;

        for (int turn = 1; turn <= length; turn++)
        {
            var ctxA = new PdCtx(lastB, turn, scoreA, scoreB, rng.NextDouble(), lastA);
            var ctxB = new PdCtx(lastA, turn, scoreB, scoreA, rng.NextDouble(), lastB);

            int rawA = DecideSafe(instA, ctxA, turnTimeoutMs, ref faultsA);
            int rawB = DecideSafe(instB, ctxB, turnTimeoutMs, ref faultsB);

            int mA = SanitizeMove(rawA).move;
            int mB = SanitizeMove(rawB).move;

            scoreA += Pd.Payoff[mA, mB];
            scoreB += Pd.Payoff[mB, mA];
            lastA = mA;
            lastB = mB;
            onTurn?.Invoke(turn, mA, mB);
        }

        string winner = scoreA > scoreB ? "A" : scoreB > scoreA ? "B" : "DRAW";
        return new MatchResult
        {
            ScoreA = scoreA, ScoreB = scoreB, Length = length,
            FaultsA = faultsA, FaultsB = faultsB, Winner = winner,
        };
    }

    // Appelle Decide avec un timeout temps-réel par tour.  En cas de dépassement ou
    // d'exception, renvoie un coup forfait (trahir) et incrémente le compteur de fautes.
    private static int DecideSafe(IPdInstance inst, PdCtx ctx, int timeoutMs, ref int faults)
    {
        if (timeoutMs <= 0)
        {
            try { return inst.Decide(ctx); }
            catch { faults++; return Pd.Defect; }
        }
        var task = Task.Run(() => inst.Decide(ctx));
        if (!task.Wait(timeoutMs))
        {
            faults++;
            return Pd.Defect; // timeout -> coup forfait
        }
        if (task.IsFaulted)
        {
            faults++;
            return Pd.Defect;
        }
        return task.Result;
    }

    /// <summary>
    /// Tournoi round-robin : chaque stratégie affronte toutes les autres ET elle-même
    /// (self-play inclus), sur les 5 longueurs.  Renvoie le classement 1er -> dernier.
    /// </summary>
    /// <param name="onPairDone">Appelé après chaque paire (i,j) avec (done, totalPairs, snapshot des stats).</param>
    public static List<StratStat> RoundRobin(
        IReadOnlyList<IPdStrategy> strategies,
        int seed,
        int reps = 1,
        Action<int, int, List<StratStat>>? onPairDone = null,
        Action<int, int, MatchResult>? onResult = null)
    {
        int n = strategies.Count;
        var stats = new List<StratStat>(n);
        for (int i = 0; i < n; i++)
        {
            var m = strategies[i].Meta;
            stats.Add(new StratStat
            {
                Index = i,
                Id = m.Id,
                Name = m.Name,
                Icon = m.Icon,
                Type = m.Type,
                IsUser = m.IsUser,
                PlayerName = m.PlayerName,
            });
        }

        var rng = new Random(seed);
        int totalPairs = n * (n + 1) / 2; // i <= j, self-play inclus
        int done = 0;

        for (int i = 0; i < n; i++)
        {
            for (int j = i; j < n; j++)
            {
                foreach (var len in Pd.Lengths)
                {
                    for (int r = 0; r < reps; r++)
                    {
                        // Graine par match pour la reproductibilité (anti-triche).
                        var res = PlayMatch(strategies[i], strategies[j], len,
                            new Random(rng.Next()));
                        stats[i].TotalScore += res.ScoreA;
                        stats[i].TotalTurns += len;
                        stats[i].Matches++;
                        stats[j].TotalScore += res.ScoreB;
                        stats[j].TotalTurns += len;
                        stats[j].Matches++;
                        onResult?.Invoke(i, j, res);
                        if (i != j)
                        {
                            if (res.ScoreA > res.ScoreB) { stats[i].Wins++; stats[j].Losses++; }
                            else if (res.ScoreA < res.ScoreB) { stats[i].Losses++; stats[j].Wins++; }
                            else { stats[i].Ties++; stats[j].Ties++; }
                        }
                    }
                }
                done++;
                onPairDone?.Invoke(done, totalPairs, stats);
            }
        }

        foreach (var s in stats)
        {
            s.AvgPerTurn = s.TotalTurns > 0 ? (double)s.TotalScore / s.TotalTurns : 0;
            s.AvgPerMatch = s.Matches > 0 ? (double)s.TotalScore / s.Matches : 0;
        }
        stats.Sort((a, b) =>
            b.AvgPerTurn.CompareTo(a.AvgPerTurn) != 0 ? b.AvgPerTurn.CompareTo(a.AvgPerTurn)
            : b.TotalScore.CompareTo(a.TotalScore));
        for (int i = 0; i < stats.Count; i++) stats[i].Rank = i + 1;
        return stats;
    }
}